newsroom/topics/topics_async.py

```diff
@@ -0,0 +1,252 @@
+    user: Annotated[Optional[ObjectIdField], validate_data_relation_async("users")] = None
+    company: Annotated[Optional[ObjectIdField], validate_data_relation_async("companies")] = None
+    is_global: bool = False
+    subscribers: Optional[List[TopicSubscriber]] = []
+    timezone_offset: Optional[int] = None
+    topic_type: TopicType
+    navigation: Optional[List[Annotated[ObjectIdField, validate_data_relation_async("navigations")]]] = None
+    folder: Annotated[Optional[ObjectIdField], validate_data_relation_async("topic_folders")] = None
+    advanced: Optional[Dict[str, Any]] = None
+
+
+class TopicService(NewshubAsyncResourceService[TopicResourceModel]):
+    async def on_create(self, docs: List[TopicResourceModel]) -> None:
+        return await super().on_create(docs)
+
+    async def on_update(self, updates: Dict[str, Any], original: TopicResourceModel) -> None:
+        await super().on_update(updates, original)
+        # If ``is_global`` has been turned off, then remove all subscribers
+        # except for the owner of the Topic
+        if original.is_global and "is_global" in updates and not updates.get("is_global"):
+            # First find the subscriber entry for the original user
+            subscriber = next(
+                (
+                    subscriber
+                    for subscriber in (updates.get("subscribers") or original.subscribers or [])
+                    if subscriber.user_id == original.user
+                ),
+                None,
+            )
+
+            # Then construct new array with either subscriber found or empty list
+            updates["subscribers"] = [subscriber] if subscriber is not None else []
+
+        if updates.get("folder"):
+            updates["folder"] = ObjectId(updates["folder"])
+
+    async def on_updated(self, updates: Dict[str, Any], original: TopicResourceModel) -> None:
+        await super().on_updated(updates, original)
+        current_user = get_user()
+
+        if current_user:
+            await auto_enable_user_emails(updates, original, current_user)
+
+    async def on_delete(self, doc: TopicResourceModel):
+        await super().on_delete(doc)
+        users = await UsersService().search(lookup={"dashboards.topic_ids": doc.id})
+        async for user in users:
+            updates = {"dashboards": user.dashboards.copy()}
+            updated_dashboards = []
+
+            for dashboard in updates["dashboards"]:
+                dashboard_dict = dashboard.to_dict()
+                # Remove the deleted topic id from topic_ids
+                dashboard_dict["topic_ids"] = [
+                    topic_id for topic_id in dashboard_dict["topic_ids"] if topic_id != doc.id
+                ]
+                updated_dashboards.append(dashboard_dict)
+
+            updates["dashboards"] = updated_dashboards
+            await UsersService().system_update(user.id, updates=updates)
```

Comment:
I'm not entirely sure about moving this logic yet. @MarkLark86 , what do you think? Would this fit better in `UsersService` or is it fine as it is?

Reply:
It should not be the Topic's responsibility to update a User when a Topic is deleted, this responsibility lies on the UserService itself. Considering we don't have an async event system yet, for a `on_topic_deleted`, maybe this can live here in Topics for now (as it is in the `develop` branch).

We can revisit this later